### PR TITLE
Adjust the entrypoint argument type of WorkgroupId dynamically

### PR DESCRIPTION
**Pull request by xuechen417**
Current the entrypoint argument type of WorkgroupId for CS is fixed as `<3xi32>`. The `MultiDispatchInfo` SGPR is assumed behind at the three SGPRs for workgroupIds. It results in a mismatch issue when registers `tgid_*_en` is optimized according to the users of workgroupId. HW will move `MultiDispatchInfo` SGPR forward if any of workgroupId component is not used. This change will dynamically adjust the argument type of workgroupId and its relevant users if partial components. We add `ShaderInput::WorkgroupId1` for using one component (`i32`) and `ShaderInput::WorkgroupId2` for using two components (`v2i32`). The process is handled by a new helper function `tryOptimizeWorkgroupId`. First, check if partial components are used. Secondly, generated corresponding `lgc.shader.input.workgroup*` call for the partial case, relink the users to the new call and remove the old call `lgc.shader.input.workgroup` and its users. Thirdly, add function attributes `amdgpu-no-workgroup-id-*` for unused components. The optimization is disabled when `m_isComputeWithCalls` is true. NOTE: we add the optimization on `tgid_*_en` in the old new pal metadata to verify the change.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit bcf6a72d6fbc472d17ca384ce876a2a335f64b09
<details><summary>CTS tests (Failed: 0/211258)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39024/70423 (55.4%)</li>
<li>Failed: 0/70423 (0.0%)</li>
<li>Not Supported: 31399/70423 (44.6%)</li><li>Warnings: 0/70423 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39058/70410 (55.5%)</li>
<li>Failed: 0/70410 (0.0%)</li>
<li>Not Supported: 31352/70410 (44.5%)</li><li>Warnings: 0/70410 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35147/70425 (49.9%)</li>
<li>Failed: 0/70425 (0.0%)</li>
<li>Not Supported: 35278/70425 (50.1%)</li><li>Warnings: 0/70425 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/patch/ShaderInputs.cpp [hunk 1]
```diff
@@ -723,3 +715,120 @@ ShaderInputs::ShaderInputUsage *ShaderInputs::getShaderInputUsage(ShaderStage st
...
+            extractInst->eraseFromParent();
+            extractInst = nullptr;
+          }
+        }
+      }
+
+      for (auto call : workgroupIdCallInsts) {
+        call->eraseFromParent();
+        call = nullptr;
+      }
```
Comment (nhaehnle) [inline lgc/patch/ShaderInputs.cpp:823]: No need to set to `nullptr`

File: lgc/patch/ShaderInputs.cpp [hunk 2]
```diff
@@ -723,3 +715,120 @@ ShaderInputs::ShaderInputUsage *ShaderInputs::getShaderInputUsage(ShaderStage st
...
+// later.
+//
+// @param pipelineState : Pipeline state
+// @param shaderStage : Shader stage
+// @param origFunc : The original entry point function
+void ShaderInputs::tryOptimizeWorkgroupId(PipelineState *pipelineState, ShaderStage shaderStage, Function *origFunc) {
+  assert(shaderStage == ShaderStageCompute);
+  bool useWorkgroupIds[3] = {false};
+  bool useWholeWorkgroupId = false;
+  SmallVector<Instruction *> extractVec3[3];
```
Comment (nhaehnle) [inline lgc/patch/ShaderInputs.cpp:733]: Avoid redundant state by removing `useWorkgroupIds` and just checking `extractVec3[i].empty()` instead where it's needed.

**Review by nhaehnle — COMMENTED** [review]
`tryOptimizeWorkgroupId` can be simplified a number of ways:

* use an early return in cases that can't be optimized, i.e. `useWholeWorkgroupId` and `usedCompCount == 3`
* you could remove `useWholeWorkgroupId` entirely using the early return
* check for gfxIp at the very beginning or even in the caller

File: lgc/patch/ShaderInputs.cpp (hunk 1 above)
Comment (xuechen417) [reply lgc/patch/ShaderInputs.cpp:823]: Fixed

File: lgc/patch/ShaderInputs.cpp (hunk 2 above)
Comment (xuechen417) [reply lgc/patch/ShaderInputs.cpp:733]: Fixed.

**Comment by xuechen417** [comment]
V1: Simplify `tryOptimizeWorkgroupId` based on the three suggestions.

**Comment by xuechen417** [comment]
V2: Fix typos

**Comment by xuechen417** [comment]
V3: Fix a logic issue resulting in lit tests failing.

**Comment by xuechen417** [comment]
> * check for gfxIp at the very beginning or even in the caller
I move add function attributes before checking for gfxip. The code piece of adding function attributes is required regardless of gfxip.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0da7bc73856de5b2df1cb2bfbe39afb216dd6d34
<details><summary>CTS tests (Failed: 0/211351)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39024/70454 (55.4%)</li>
<li>Failed: 0/70454 (0.0%)</li>
<li>Not Supported: 31430/70454 (44.6%)</li><li>Warnings: 0/70454 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39058/70441 (55.4%)</li>
<li>Failed: 0/70441 (0.0%)</li>
<li>Not Supported: 31383/70441 (44.6%)</li><li>Warnings: 0/70441 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35147/70456 (49.9%)</li>
<li>Failed: 0/70456 (0.0%)</li>
<li>Not Supported: 35309/70456 (50.1%)</li><li>Warnings: 0/70456 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — COMMENTED** [review]
I think that the attribute-adding logic in `fixupUses` should just be removed: if we still use `ShaderInput::WorkgroupId` (instead of WorkgroupId1 or 2), then this means that `tryOptimizeWorkgroupId` wasn't successful in removing a component. But then why should `fixupUses` be successful?

Apart from that, this is looking much better already. Thank you!

File: lgc/patch/ShaderInputs.cpp [hunk 3]
```diff
@@ -335,6 +343,7 @@ void ShaderInputs::fixupUses(Module &module, PipelineState *pipelineState) {
     // Use for compute shader
     bool useWorkgroupIds[3] = {false};
     bool useWholeWorkgroupId = false;
+    bool needCheckWorgroupId = false;
```
[the remark this answers is not in the document]
Comment (xuechen417) [reply lgc/patch/ShaderInputs.cpp:346]: Thanks. Fixed.

**Comment by xuechen417** [comment]
> I think that the attribute-adding logic in `fixupUses` should just be removed: if we still use `ShaderInput::WorkgroupId` (instead of WorkgroupId1 or 2), then this means that `tryOptimizeWorkgroupId` wasn't successful in removing a component. But then why should `fixupUses` be successful?
> 
I replied in the email. Please check it.

**Comment by xuechen417** [comment]
V5: rebase and fix typo

**Comment by amdvlk-admin** [comment]
### Test summary for commit 69cfb207d9342452f1fe7ac4f4e9f5d3827ec586
<details><summary>CTS tests (Failed: 0/211351)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39024/70454 (55.4%)</li>
<li>Failed: 0/70454 (0.0%)</li>
<li>Not Supported: 31430/70454 (44.6%)</li><li>Warnings: 0/70454 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39058/70441 (55.4%)</li>
<li>Failed: 0/70441 (0.0%)</li>
<li>Not Supported: 31383/70441 (44.6%)</li><li>Warnings: 0/70441 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35147/70456 (49.9%)</li>
<li>Failed: 0/70456 (0.0%)</li>
<li>Not Supported: 35309/70456 (50.1%)</li><li>Warnings: 0/70456 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**